### ISSUE TITLE
Node.js: Evaporate requires support for File and Promise

### ISSUE THREAD

**Issue by AJB99** [TTLabs/EvaporateJS#391]
Hi there,

I'm attempting to use EvaporateJS in my node.js 6.6.0 application and I'm getting the following error when I run my app:

`UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 2): Evaporate requires support for File and Promise`

Can EvaporateJS not be used as a node module on the server-side?
Is it a requirement that it is run in the browser?

**Comment by bikeath1337** [comment]
Yes, as the description of the project describes. The readme also references node.js and there is more information in the wiki. However, in summary,

if you want to use in Node, then you need to include Node's file system. If your environment does not support a native Promise (see the docs), then you need to include that as well and all should work. There is an example as well that covers an Electron integration which you can refer to.

**Comment by AJB99** [comment]
Hi @bikeath1337, thanks for the quick reply. Still need a little clarification if you could ...

`Yes, as the description of the project describes.`
So, the answer is that yes, EvaporateJS *can* be used as a node.js module?

`if you want to use in Node, then you need to include Node's file system.`
Are you referring to `var fs = require('fs)`? That's already being done, but the error persists.

`If your environment does not support a native Promise (see the docs), then you need to include that as well and all should work.`
I'm running node 6.6.0, so Promises are natively supported (since 0.12.0) so far as I understand. But I'm still getting the error.

Maybe it's best if I provide an example of my setup:

```
var fs = require('fs');
var crypto = require('crypto');
var evaporate = require('evaporate');


var fileBuffer = fs.readFileSync('./.test/media/image-1.jpg');

evaporate.create({
	aws_key : process.env.AWS_ACCESS_KEY_ID,
	bucket : process.env.AWS_S3_CDN_BUCKET,
	cloudfront : true,
	computeContentMd5 : true,
	cryptoMd5Method : function(data) {
		return crypto.createHash('md5').update(data).digest('base64');
	},
	customAuthMethod : function(signParams, signHeaders, stringToSign, signatureDateTime, canonicalRequest) {

		console.log(arguments);

		// return new Promise(function (resolve, reject) {
		// 	resolve('computed signature');
		// }

	        }
}).then(function(evap) {

	evap.add({
		name : 'image-1.jpg',
		file : fileBuffer,
		progress : function(progressValue) {
			console.log('Progress', progressValue);
		},
		complete: function(_xhr, awsKey) {
			console.log('Complete!');
		}
	},{
		bucket : AWS_BUCKET // Shows that the bucket can be changed per
	})
	.then(function(awsObjectKey) {
		console.log('File successfully uploaded to:', awsObjectKey);
	},
	function(reason) {
		console.log('File did not upload sucessfully:', reason);
	});

});

```

**Comment by AJB99** [comment]
Hi @bikeath1337 Is it possible that you could elaborate on the above. Thanks in advance for your time.

**Comment by bikeath1337** [comment]
I am not a node developer but the intention was with v2.0 to remove dependence on a browser. At that time, I had confirmation that browsers were not supported and it was a feature that seemed to make the community happy. The reference to File may have been overlooked because most of the focus was on Electron support, which happens to include an implementation of `File`.

In any case, it is possible that the guard you're triggering is no longer necessary. Why don't you remove the guard check for `File` and see if that solves your problem? Please not that you need to provide the FS stream callback method following samples in the docs/examples `readableStreamPartMethod`.

I would like confirmation from Node developers regarding removing that guard and would welcome a PR if it's no longer needed.

**Comment by jakubzitny** [comment]
@AJB99: You need to provide a `File` constructor to environment (`global.File`) in Node to bypass Evaporate's check for it. It is not really needed, but if you do that, and properly include `readableStreamPartMethod`, it should work. You can use `File` from [file-api](https://www.npmjs.com/package/file-api) module to mock the File environment. I do it in tests like that.

@bikeath1337 yeah, the guard could be improved. I'd not remove it though, rather, I'd improve it so it checks both scenarios properly.

**Comment by Warry** [comment]
I might also add that it could help to have a Promise/callback in the browser as well, for exemple to compute the crypto hashs inside a webworker.

**Comment by masonlee** [comment]
Thanks guys for this great project!  I was able to get Evaporate running in pure node (not in the browser) with the following code. Thought I would share here.  A little more than just `File` is needed to pass all the checks and satisfy all the runtime requirements. :)  (cc: @bikeath1337 )

## Running Evaporate in Node.js (8.9.4) ## 
```
/* The following is the extra setup code you need in order to run evaporate in pure node,
plus a utility function showing how to make an HTML5 File object that Evaporate will like.  */

const { URL } = require('url');
const XMLHttpRequest = require('xhr2'); // https://github.com/pwnall/node-xhr2
const { File } = require('file-api'); // https://github.com/node-file-api/file-api

// global hacks to satisfy all evaporate's environment assumptions..
global.URL = URL;
global.XMLHttpRequest = XMLHttpRequest;
global.File = File;
global.FileReader = function FileReader() {};
global.FileReader.prototype.readAsArrayBuffer = () => {};

// Note that Evaporate won't work with streams alone-- it still needs an HTML5 File object,
// which it uses to get the file size and passes to some of the callbacks.
// file-api (imported above) seems to implement the File api well enough.
function createHTML5File(filePath) {
  return new File({
    path: filePath,
    type: 'application/octet-stream', // Need something here or file-api fails on mime lookup
    async: false,
    jsdom: true, // We need this option to have the file size preloaded.
  });
}
```
